From df1ee070091f0cb9775854f7e1075ca94f023cd9 Mon Sep 17 00:00:00 2001
From: Yuantian Tang <andy.tang@nxp.com>
Date: Fri, 18 Aug 2017 14:52:52 +0800
Subject: [PATCH 065/741] drivers: firmware: psci: use psci v0.2 to implement
 sleep

commit 186f11e5ff670a4ab8c7fbc7efdd829b0c0996db from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Specification PSCI v0.2 doesn't define function SYSTEM_SUSPEND
but v1.0 do. However, so far PPA only support PSCI v0.2 as well
as SYSTEM_SUSPEND function. So a temp workaround is needed to
hack PSCI driver to enable suspend call even in case of PSCI v0.2

Signed-off-by: Tang Yuantian <andy.tang@nxp.com>
Signed-off-by: Yinbo Zhu <yinbo.zhu@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/firmware/psci/psci.c | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/drivers/firmware/psci/psci.c b/drivers/firmware/psci/psci.c
index f82ccd39a913..d8f304f4ef7d 100644
--- a/drivers/firmware/psci/psci.c
+++ b/drivers/firmware/psci/psci.c
@@ -450,8 +450,20 @@ CPUIDLE_METHOD_OF_DECLARE(psci, "psci", &psci_cpuidle_ops);
 
 static int psci_system_suspend(unsigned long unused)
 {
-	return invoke_psci_fn(PSCI_FN_NATIVE(1_0, SYSTEM_SUSPEND),
-			      __pa_symbol(cpu_resume), 0, 0);
+
+        u32 state;
+        u32 ver = psci_get_version();
+ 
+        if (PSCI_VERSION_MAJOR(ver) >= 1) {
+                return invoke_psci_fn(PSCI_FN_NATIVE(1_0, SYSTEM_SUSPEND),
+                                virt_to_phys(cpu_resume), 0, 0);
+        } else {
+                state = (2 << PSCI_0_2_POWER_STATE_AFFL_SHIFT) |
+                        (1 << PSCI_0_2_POWER_STATE_TYPE_SHIFT);
+ 
+                return psci_cpu_suspend(state, virt_to_phys(cpu_resume));
+        }
+
 }
 
 static int psci_system_suspend_enter(suspend_state_t state)
@@ -585,6 +597,8 @@ static void __init psci_0_2_set_functions(void)
 	arm_pm_restart = psci_sys_reset;
 
 	pm_power_off = psci_sys_poweroff;
+
+	suspend_set_ops(&psci_suspend_ops);
 }
 
 /*
-- 
2.17.1

